### PR TITLE
Extend approval options

### PR DESCRIPTION
**Pull request by anbraten**
related to #336

To protect secrets and (agents) it should be possible to require approvals for potential harmful pipelines like PRs. To not require every pipeline to be approved the existing gated mode was replaced with 3 options addressing most common cases:

- allow every pipeline except PRs from forks
- require approval for all PRs (new default)
- require approval for every pipeline (current require approval / gated mode)

Those modes could later on be extended. Additionally a webhook extension could be added to get the approval state from an external service.

# Changes
- [x] replace `IsGated` with `RequireApproval`
- [x] migrate old settings to `RequireApproval`
- [x] adjust docs

### CODE REVIEW

**Comment by xoxys** [comment]
> Can't you create an interface which just checks whether the pipeline needs approval or not so we can easily extend the feature by adding a new implementation? Maybe also with support for an external http extension?

If we need that level of flexibility, why not use something like [open policy agent](https://www.openpolicyagent.org/), as proposed in the issue? Repo rules can be written by users and added via UI by repo admins without writing external services and without blowing up the server code for every single use case that we have not yet thought about.

**Comment by anbraten** [comment]
> In my opinion, we should support more here.

Sure, we can make it highly adjustable, but on the other hand, to keep things simple (which is often a quite import benefit / feature as well for the community) it's sometimes more helpful to support just the most common options IMO.

> If we need that level of flexibility, why not use something like [open policy agent](https://www.openpolicyagent.org/), as proposed in the issue? Repo rules can be written by users and added via UI by repo admins without writing external services and without blowing up the server code for every single use case that we have not yet thought about.

I don't think [open policy agent](https://www.openpolicyagent.org/) would be that helpful as we still need to provide context data (usernames, if user is admin / maintainer / ..., amount of pipeline executed for a context (like PR), ...) to be even able to write rules 
 as suggested here: https://github.com/woodpecker-ci/woodpecker/discussions/2293#discussioncomment-6774908

**Comment by xoxys** [comment]
Sure, but implementing something like a `SecurityContext` would be required for external services as well.

> it's sometimes more helpful to support just the most common options IMO.

To be honest (please don't get me wrong) we also implement stuff like cli auto-updates cli auto-setup etc. just because it's convenient. This also adds code complexity, and we still do it even if it is easier to just let users download the binary from a release. Especially for security related features, flexibility is key IMO.

**Comment by 6543** [comment]
i still want to have the gated feature to be extendable

**Comment by xoxys** [comment]
Yes, please... if we dont want to have more options in woodpecker core (I still dont understand why) please provide an http interface like drone ci does so users can implement validation.

**Comment by lafriks** [comment]
There should be an option to not require approval from forks as this would be quite problematic for company internal instances imho

**Comment by xoxys** [comment]
> There should be an option to not require approval from forks as this would be quite problematic for company internal instances imho

Fully agree. Otherwise, this PR is also breaking, as right now approvals can be disabled entirely.

File: cli/repo/repo_update.go
Comment (xoxys) [inline cli/repo/repo_update.go:44]:
```suggestion
			Usage: "repository requires approval",
```

Comment (xoxys) [reply cli/repo/repo_update.go:44]: Or

```suggestion
			Usage: "repositories require approval",
```

File: cli/repo/repo_update.go
Comment (xoxys) [inline cli/repo/repo_update.go:107]: Can we use constants for these strings?

Comment (xoxys) [reply cli/repo/repo_update.go:107]: Ive seen there is already `RequireApprovalAllEvents` etc. can we not use it here as well?

**Comment by anbraten** [comment]
> There should be an option to not require approval from forks as this would be quite problematic for company internal instances imho

Was thinking about this. So you basically have one fork per employee?

File: cli/repo/repo_update.go
Comment (anbraten) [reply cli/repo/repo_update.go]: I just copied the types from server/models to woodpecker-go now.

File: docs/docs/20-usage/project-settings.png
Comment (xoxys) [inline docs/docs/20-usage/project-settings.png:1]: On the screenshot is no radio button selected. Shouldn't "No approval required" selected by default? IMO, we should change the default in the upcoming v3 to "Pull request from forks" then.

Comment (anbraten) [reply docs/docs/20-usage/project-settings.png:1]: Updated it. Used `ci.woodpecker-ci.org` as api for the UI and that wasn't returning any value for the new setting 😅 

Currently it sets the level to `require approval for all pull-requests` for new repos and migrates existing ones like this:
- gated => still require approval for everything
- prs enabled flag checked => require approval for all prs
- else => require approval for forks (if someone as lafriks mentioned wants to completely disable approvals they would need to explicitly change this)

Should I change some part of this?

File: cli/repo/repo_update.go
Comment (xoxys) [inline cli/repo/repo_update.go:100]: Changing the behavior for `gated == false` is also breaking and should be avoided if this PR should go into 2.7.x

Comment (xoxys) [reply cli/repo/repo_update.go:100]: Lets discuss in the other thread.

File: docs/docs/20-usage/project-settings.png
Comment (xoxys) [reply docs/docs/20-usage/project-settings.png:1]: For existing repos we should not change the behavior in v2.7.x

> * gated => still require approval for everything

That's fine

> * prs enabled flag checked => require approval for all prs

That's breaking, we can and should do it that way but not in v2.7.x

> * else => require approval for forks (if someone as lafriks mentioned wants to completely disable approvals they would need to explicitly change this)

Same here.

**Comment by pat-s** [comment]
The milestone was set to 2.7.2 - @anbraten I think this should be 3.0?

WRT to backporting to 2.x: is this really needed? This is quite a big feature and if at all, this should go into a point release but not a patch release. I see the tendency lately to start backporting more into 2.x and keeping that alive/around longer - is that the idea?

**Comment by anbraten** [comment]
@pat-s I believe we should release this as soon as possible as a robust measure to address security risks. We’ve implemented multiple fixes recently - #3934, #4076, #3933, #3909, #4053, and #3918, to name a few - each balancing security, UX, and potential breaking changes. 

While researching how other CI systems manage cases like the potentially "risky" `PATH` environment variable 🙈, it’s clear they often avoid such restrictions entirely. The rationale? Strict limitations can hinder legitimate use cases, and restricting these features without impacting users is challenging. Consequently, they largely rely on not running risky pipelines in the first place - a pragmatic approach IMO.

The only reliable solution is to avoid executing potentially dangerous code from untrusted sources without prior review. To that end, I'd like to adjust our security approach to incorporate this chain of trust, addressing the risks outlined [here](https://github.com/woodpecker-ci/woodpecker/discussions/4234#discussioncomment-10990275):

- **Repo admins**: Ensure only "safe" code is executed. If trust is in question, conduct a review.
- **Agent admins**: Choose which repos/orgs to trust, and limit features to those within your control.

**Comment by lafriks** [comment]
> > There should be an option to not require approval from forks as this would be quite problematic for company internal instances imho
> 
> Was thinking about this. So you basically have one fork per employee?

Mostly we have product teams but there could be coming fixes from other teams also and in such cases it would be fork

File: server/model/repo.go
Comment (6543) [inline server/model/repo.go:29]:
```suggestion
	RequireApprovalAllEvents    ApprovalMode = "all_events"    // require approval for all external events
```

File: web/src/assets/locales/en.json
Comment (6543) [inline web/src/assets/locales/en.json:499]: all events description: all events that are external, so cron and manual events will run right away ...

**Comment by 6543** [comment]
> The milestone was set to 2.7.2 - @anbraten I think this should be 3.0?
> 
> WRT to backporting to 2.x: is this really needed? This is quite a big feature and if at all, this should go into a point release but not a patch release. I see the tendency lately to start backporting more into 2.x and keeping that alive/around longer - is that the idea?
> 
> Besides: great feature, thanks a lot for that! It will make administrating public instances a lot easier!

this can never be v2.7.x ... and we hold back v2.7.2 since for toooo long ...

If this should be released asap too ... just let us make a v2.8 release branch after v2.7.2 and continue on that ... also we could cut out v3 soonish - we wont get everything anyway .... even things I want to see, so just stick to do that into a new cycle ...

**Comment by 6543** [comment]
I still think we could add https://www.openpolicyagent.org/ or something alike .... but the current implementation is already better, and for better UX having simple pre-selected options is a good choice in any case

File: docs/docs/20-usage/project-settings.png
Comment (anbraten) [reply docs/docs/20-usage/project-settings.png:1]: I adjusted the logic when changing using the gated property via api to map `no gated` to `no approval required`. 

For the migration I changed it to:
- gated => approval for all events
- public repo => approval for forks (I would argue this is fine as the option was not available before, although everyone would have chosen it)
- private / internal repo & not gated => no approval required (this should allow teams as @lafriks mentioned to use internal forks without approval)

File: server/model/repo.go
Comment (6543) [inline server/model/repo.go:53]: fmt

File: server/model/repo.go
Comment (6543) [inline server/model/repo.go:121]:
```suggestion

```

**Comment by 6543** [comment]
It can be v2.8.0 ... once v2.7.2 got released ...

**Comment by pat-s** [comment]
2.7.2 can just be a 2.8.0 then. Doesn't make much sense to chain a 2.8.0 shortly after 2.7.2?

File: server/model/repo.go
Comment (anbraten) [reply server/model/repo.go]: adjusted

**Review by 6543 — APPROVED** [review]

**Comment by 6543** [comment]
@anbraten I updated the migration, please have a look at it and :eyes:  :)

**Comment by xoxys** [comment]
Looks like you broke the tests now

File: server/store/datastore/migration/017_gated_to_require_approval.go
Comment (6543) [inline server/store/datastore/migration/017_gated_to_require_approval.go:49]: Me: from forks to pullrequest, as it is repo default on repo creation ... throug forks is also fine by me ...

@anbraten if you have a strong opinion just change it to forks

(It got updated by me while i rewrote the migration to not require pagination)

**Comment by xoxys** [comment]
I'm confused about the milestone. Is this one targeted for v3 now or 2.8? And if the target for v3.0 is correct, why is there a `backport-required` label also present?

**Comment by 6543** [comment]
@anbraten I think you can merge it any time now :)

File: docs/docs/20-usage/75-project-settings.md
Comment (lafriks) [inline docs/docs/20-usage/75-project-settings.md:30]: This should be updated for no approval required mode, should probably update to say that default and recommended mode is to require approval from forks

**Comment by 6543** [comment]
conflicts :/

File: server/store/datastore/migration/017_gated_to_require_approval.go
Comment (6543) [reply server/store/datastore/migration/017_gated_to_require_approval.go]: reverted the thing back to forks -> f54e288cc6fbdd0004c17a06f80eaa4463c387f1

File: docs/docs/20-usage/75-project-settings.md
Comment (6543) [reply docs/docs/20-usage/75-project-settings.md]: b17ce97da535553fa9fc4d8e0145eee5efd55817

**Review by 6543 — APPROVED** [review]

**Comment by 6543** [comment]
please backport :)

**Comment by xoxys** [comment]
> please backport :)

This is IMO pretty bad practice. If we merge the PR from someone else, we can't expect that the other one will be able to create the backport soon.

**Comment by xoxys** [comment]
@6543 See discussion in https://github.com/woodpecker-ci/woodpecker/pull/3348#discussion_r1826043277

Migrating a public, not gated repo to RequiredForkApproval is by definition a breaking change as it changes the behavior for users. Yes, you can argue that for security reasons we accept this tradeoff, but then this should get a pretty clear note in the release notes and proper documentation.

**Pull request merged**
